File: app/javascript/channels/message_channel.js
import consumer from "./consumer"

consumer.subscriptions.create("MessageChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
    console.log("Connected to the message channel.")
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
    console.log("Disconnected from the message channel.")
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
    const messageContainer = document.getElementById('message-container');
    const messageElement = document.createElement('p');
    messageElement.textContent = `${data.user}: ${data.content}`;
    messageContainer.appendChild(messageElement);
  },
  speak(message, roomId) {
    this.perform('speak', { message: message, room_id: roomId })
  }
});

File: app/channels/message_channel.rb
class MessageChannel < ApplicationCable::Channel
  def subscribed
    # stream_from "some_channel"
    stream_from "message_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end
  def speak(data)
    Message.create(content: data['message'], room_id: data['room_id'], user_id: current_user.id)
  end
end
